File: Roadmap/04 - CADENAS DE CARACTERES/python/D3rk1us.py
"""
 * EJERCICIO:
 * Muestra ejemplos de todas las operaciones que puedes realizar con cadenas de caracteres
 * en tu lenguaje. Algunas de esas operaciones podrían ser (busca todas las que puedas):
 * - Acceso a caracteres específicos, subcadenas, longitud, concatenación, repetición, recorrido,
 *   conversión a mayúsculas y minúsculas, reemplazo, división, unión, interpolación, verificación...
 *
 * DIFICULTAD EXTRA (opcional):
 * Crea un programa que analice dos palabras diferentes y realice comprobaciones
 * para descubrir si son:
 * - Palíndromos
 * - Anagramas
 * - Isogramas
 """

nombre_cadena = 'Alfabeto'
cadena = "abcd-efg-hijk"

 # Acceso a caracteres específicos

print(cadena[1])

 # Subcadenas

print(cadena[5:8])

 # Longitud

print(len(cadena))

 # Concatenación

print(nombre_cadena + ":" + cadena)
print(nombre_cadena, ":", cadena)
print(f"{nombre_cadena}: {cadena}")

 # Repetición 

print(cadena * 4)

 # Recorrido

for c in cadena:
    print(c)

 # Conversión a mayúsculas y minúsculas

print(cadena.upper())
print(nombre_cadena.lower())

 # Reemplazo

print(cadena.replace('a', 'z'))

 # División y Unión

cadena = cadena.split('-')
print(cadena)

cadena = ' '.join(cadena)
print(cadena)

 # Interpolación

print(f"Este es el {nombre_cadena.lower()}, que comienza con: {cadena}")

 # Verificación

print(isinstance(cadena, str))
print(type(nombre_cadena))

"""
/////////////////////////////////////////////////  DIFICULTAD EXTRA  /////////////////////////////////////////////////
"""

palabra1 = "adan"
palabra2 = "nada"

# Funciones que comprueban si son palíndromos, anagramas y/o isogramas

def palindromo(palabra1, palabra2):
    list_palabra1 = [*palabra1] # Divide cada caracter en un elemento de un listado.
    list_palabra2 = [*palabra2]

    list_palabra2.reverse()

    if list_palabra1 == list_palabra2:
        return "Palíndromos."
    else:
        return "No son palíndromos."

def anagrama(palabra1, palabra2):
    list_palabra1 = [*palabra1] # Divide cada caracter en un elemento de un listado.
    list_palabra2 = [*palabra2]

    list_palabra1.sort()
    list_palabra2.sort()

    if list_palabra1 == list_palabra2:
        return "Anagramas."
    else:
        return "No son anagramas."

def isograma(palabra):
    count = 0 # Cuenta cuánto se repite una letra.

    for a in range(len(palabra)):
        for k in palabra:

            if palabra[a] == k:
                count += 1
                if count == 2:  #En el momento que una letra se repite dos veces, deja de ser Isograma.
                    return "No es Isograma."
        
        count = 0
    
    return "Es Isograma."

print(palindromo(palabra1, palabra2))
print(anagrama(palabra1, palabra2))
print(isograma(palabra2))

File: Roadmap/04 - CADENAS DE CARACTERES/dart/D3rk1us.dart
/*
 * EJERCICIO:
 * Muestra ejemplos de todas las operaciones que puedes realizar con cadenas de caracteres
 * en tu lenguaje. Algunas de esas operaciones podrían ser (busca todas las que puedas):
 * - Acceso a caracteres específicos, subcadenas, longitud, concatenación, repetición, recorrido,
 *   conversión a mayúsculas y minúsculas, reemplazo, división, unión, interpolación, verificación...
 */

void main() {

  String cadena1 = 'Cadena con comillas simples.';
  String cadena2 = "Cadena con comillas dobles.";
  String cadena3 = '''
  Cadena
  multilinea 
  de tres comillas simples''';

  print(cadena1);
  print(cadena2);
  print(cadena3);

  // Subcadenas
  print(cadena1.substring(0, 10));

  // Longitud
  print(cadena2.length);

  // Concatenación
  print("Concatenación: " + cadena2); 

  // Repetición
  for (int i = 0; i <= 2; i++) {
    print('Repetición');
  }

  // Recorrido
  for (var c = 0; c < cadena2.length; c++) {
    print(cadena2[c]);
  }
  
  // Conversión a mayúsculas y minúsculas
  print(cadena1.toUpperCase());
  print(cadena3.toLowerCase());

  // Reemplazo
  print(cadena2.replaceAll('Cadena', 'Ejemplo'));

  // División
  print(cadena1.split(' '));

  // Unión
  print(cadena1.split(' ').join('-'));

  // Interpolación
  print("Esto es una $cadena2");

  // Verificación
  print(cadena1 == cadena2);
  print(cadena1.compareTo(cadena3));

  /*
  * DIFICULTAD EXTRA (opcional):
  * Crea un programa que analice dos palabras diferentes y realice comprobaciones
  * para descubrir si son:
  * - Palíndromos
  * - Anagramas
  * - Isogramas
  */

  String palabra1 = 'ana';
  String palabra2 = 'amor';

  print('$palabra1: ${palindromo(palabra1)}');
  print('$palabra2: ${palindromo(palabra2)}');

  print('$palabra1 y $palabra2: ${anagrama(palabra1, palabra2)}');

  print('$palabra1: ${isograma(palabra1)}');
  print('$palabra2: ${isograma(palabra2)}');

}

String palindromo(String palabra) {
  
  var palPalabra = palabra.split('').reversed.join();
  
  if (palabra == palPalabra) {
    return 'Es un palíndromo';
  } else {
    return 'No es palíndromo';
  }
}

String anagrama(String palabra1, String palabra2) {
  
  var anPalabra1 = palabra1.split('');
  anPalabra1.sort();
  
  var anPalabra2 = palabra2.split('');
  anPalabra2.sort();
 
  if (anPalabra1.join() == anPalabra2.join()) {
    return 'Son anagramas.';
  } else {
    return 'No son anagramas.';
  }
}

String isograma(String palabra) {
  int count = 0;
  var comprobar = palabra.split('');

  for (int i = 0; i < comprobar.length; i++) {

    for (var caracter in comprobar) {
      if (caracter == palabra[i]) {
        count += 1;
        if (count == 2) {
          return 'No es Isograma';
        }
      }
    }
    count = 0;
  }
  return 'Es un Isograma.';
}